release/internal/hashreleaseserver/ssh.go

```diff
@@ -28,81 +30,68 @@ import (
 	"github.com/skeema/knownhosts"
 )
 
-// SSHConfig holds the configuration for an SSH connection
-type SSHConfig struct {
-	Host    string
-	User    string
-	KeyPath string
-	Port    string
-}
-
-// NewSSHConfig creates a new SSHConfig
-func NewSSHConfig(host, user, keyPath, port string) *SSHConfig {
-	return &SSHConfig{
-		Host:    host,
-		User:    user,
-		KeyPath: keyPath,
-		Port:    port,
-	}
-}
-
-// Args returns the ssh command string arguments
-func (s *SSHConfig) Args() string {
-	str := []string{"-i", s.KeyPath, "-p", s.Port, "-q", "-o StrictHostKeyChecking=no", "-o UserKnownHostsFile=/dev/null"}
-	return strings.Join(str, " ")
-}
-
-// HostString returns the host string in the format user@host
-func (s *SSHConfig) HostString() string {
-	return s.User + "@" + s.Host
-}
-
-// Address returns the address in the format host:port
-func (s *SSHConfig) Address() string {
-	return fmt.Sprintf("%s:%s", s.Host, s.Port)
-}
-
-func connect(sshConfig *SSHConfig) (*ssh.Session, error) {
-	key, err := os.ReadFile(sshConfig.KeyPath)
+func connect(cfg *Config) (*ssh.Session, error) {
+	key, err := os.ReadFile(cfg.Key)
 	if err != nil {
-		logrus.WithField("key", sshConfig.KeyPath).WithError(err).Error("Unable to read ssh key")
+		logrus.WithField("key", cfg.Key).WithError(err).Error("Unable to read ssh key")
 		return nil, err
 	}
 	signer, err := ssh.ParsePrivateKey(key)
 	if err != nil {
 		return nil, err
 	}
 	config := &ssh.ClientConfig{
-		User: sshConfig.User,
+		User: cfg.User,
 		Auth: []ssh.AuthMethod{
 			ssh.PublicKeys(signer),
 		},
-		// This callback mimics the behavior of ssh -o StrictHostKeyChecking=no
-		HostKeyCallback: ssh.HostKeyCallback(func(host string, remote net.Addr, pubKey ssh.PublicKey) error {
+		HostKeyCallback: ssh.HostKeyCallback(func(host string, remote net.Addr, key ssh.PublicKey) error {
 			knownHostsFilePath := filepath.Join(os.Getenv("HOME"), ".ssh", "known_hosts")
 			k, err := knownhosts.NewDB(knownHostsFilePath)
 			if err != nil {
 				return err
 			}
-			err = k.HostKeyCallback()(host, remote, pubKey)
+			err = k.HostKeyCallback()(host, remote, key)
 			if knownhosts.IsHostKeyChanged(err) {
 				return fmt.Errorf("host key changed: %v", err)
 			} else if knownhosts.IsHostUnknown(err) {
-				f, err := os.OpenFile(knownHostsFilePath, os.O_APPEND|os.O_WRONLY, 0o600)
-				if err != nil {
-					return err
-				}
-				defer f.Close()
-				err = knownhosts.WriteKnownHost(f, host, remote, pubKey)
-				if err != nil {
-					return err
+				// When HostKeyCallback returns an error with IsHostUnknown,
```

Comment:
I'm not sure this logic really belongs here - generally you don't want to just blindly add entries to the known_hosts file. Rather, you should check the host key is correct before adding it.

I would expect a better approach to be that our CI systems are pre-configured with a known_hosts entry for this. Is that feasible?

@Behnam-Shobiri WDYT?

Reply:
Just to explain this logic:

- Check if the public key from the server is available in `known_hosts` file (classic when verifying keys)
- If it is not available but the user has the public key saved somewhere and specifies it (using `DOCS_HOST_KEY`)
  - if the `DOCS_HOST_KEY` matches the public key from the server then it attempt to add it to `known_hosts` and proceed with connection
  - if the `DOCS_HOST_KEY` matches but cannot be added to `known_hosts` file, the host is still valid so proceed with connection
  - if the `DOCS_HOST_KEY` does not match it errors out

I went with this approah as I felt it will ease rotating the key. I think what you are suggesting _might_ require the `known_hosts` file to be overwritten _potentially_ in CI via secrets and I was very hesistant about that.

Reply:
@radTuti I understand what you trying to do but it seems a bit complicated. It wouldn't be easier if we appended the host key to the known host file at the beginning and then used ssh.[FixedHostKey](https://pkg.go.dev/golang.org/x/crypto/ssh#FixedHostKey) function? 

The comparison seems fine but it may increase the attack surface since it is not the best way to compare keys as strings. I don't know about what attack might we find now or in the future (we might be subject to some injection attacks); however, we have an easier solution. Let me know if I'm missing something.